### PR TITLE
build: run tests on python 3.11

### LINKED ISSUE #4307: [course-discovery] Test Python 3.11 and 3.12
We should ideally aim for python 3.12 but if there are major issues, we should fall back to 3.11

### PR DESCRIPTION
**Pull request by iamsobanjaved**

### CODE REVIEW

**Review by UsamaSadiq — APPROVED** [review]

**Comment by e0d** [comment]
@iamsobanjaved  Your branch is behind the base.  I've pulled in changes from master as a merge commit which will update your branch and cause the tests to be re-run. 

I did notice that the codecoverage check failed on the prior build.

**Comment by DawoudSheraz** [comment]
@iamsobanjaved Please update PR description, thank you.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -54,12 +70,16 @@ jobs:
...
+          PYTHON_VERSION: 3.8
       - run: make ci_quality
 
   semgrep:
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v3
       - run: make ci_up
+        env:
+          PYTHON_VERSION: 3.8
```
Comment (DawoudSheraz) [inline .github/workflows/ci.yml]: shouldn't semgrep be run across all versions?

File: Dockerfile [hunk 2]
```diff
@@ -1,20 +1,27 @@
...
   language-pack-en \
   build-essential \
-  python3.8-dev \
-  python3-virtualenv \
-  python3.8-distutils \
+  python${PYTHON_VERSION}-dev \
+  python${PYTHON_VERSION}-distutils \
   libmysqlclient-dev \
   libssl-dev \
+  # Current version of Pillow (9.5.0) doesn't provide pre-built wheel for python 3.12,
```
Comment (DawoudSheraz) [inline Dockerfile]: nit: add TODO: at the beginning.

File: Dockerfile [hunk 3]
```diff
@@ -38,8 +45,12 @@ ENV PATH "${DISCOVERY_VENV_DIR}/bin:${DISCOVERY_NODEENV_DIR}/bin:$PATH"
 ENV DISCOVERY_CFG "/edx/etc/discovery.yml"
 ENV DISCOVERY_CODE_DIR "${DISCOVERY_CODE_DIR}"
 ENV DISCOVERY_APP_DIR "${DISCOVERY_APP_DIR}"
+ENV PYTHON_VERSION "${PYTHON_VERSION}"
+
+RUN curl -sS https://bootstrap.pypa.io/get-pip.py | python${PYTHON_VERSION}
+RUN pip install virtualenv
 
-RUN virtualenv -p python3.8 --always-copy ${DISCOVERY_VENV_DIR}
+RUN virtualenv -p python${PYTHON_VERSION} --always-copy ${DISCOVERY_VENV_DIR}
```
Comment (DawoudSheraz) [inline Dockerfile:53]: nit: isn't venv recommended for Py3? Just wondering if this be venv instead of virtualenv.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -18,6 +18,20 @@ jobs:
...
+        run: |
+          # Remove 'py' and insert a dot to format the version
+          FORMATTED_VERSION=${{ matrix.python-version }}  # e.g., py38
+          FORMATTED_VERSION=${FORMATTED_VERSION/py3/3.}    # becomes 3.8
+          
+          # Set environment variables
+          echo "PYTHON_VERSION=$FORMATTED_VERSION" >> $GITHUB_ENV
+          
+          # Output formatted version for use in subsequent steps
+          echo "Using Python Version: $PYTHON_VERSION"
```
Comment (zawan-ila) [inline .github/workflows/ci.yml]: [nit] PYTHON_VERSION is still undefined here. 
From the docs
`The step that creates or updates the environment variable does not have access to the new value, but all subsequent steps in a job will have access `

**Review by zawan-ila — APPROVED** [review]

File: Dockerfile (hunk 2 above)
Comment (feanil) [inline Dockerfile]:
```suggestion
  # TODO: Current version of Pillow (9.5.0) doesn't provide pre-built wheel for python 3.12,
```

File: .github/workflows/ci.yml (hunk 4 above)
Comment (feanil) [inline .github/workflows/ci.yml]:
```suggestion
          echo "Using Python Version: $FORMATTED_VERSION"
```

**Review by feanil — APPROVED** [review]

File: Dockerfile (hunk 3 above)
Comment (feanil) [reply Dockerfile:53]: This is fine, both should work.

**Pull request merged**